Laser pointer update: Shining Through Walls Edition (feat. fixes!)

Sealed101 opened the pull request:
# _PR PSA_

![augh](https://github.com/tgstation/tgstation/assets/75863639/6dc87fc7-65a3-4b7c-9b8d-a1432cacbe93)


## About The Pull Request
Cleans up code for laser pointers, fixing some bugs like the forever-charging state or affecting dead cats along the way.
Remaining charge is now available upon examine.
Canonizes #45834 by implementing an upgrade to the laser pointers: installing a bluespace crystal into a laser with tier 3 or higher laser diode lets it shine through walls. Using an upgraded laser uses twice the charge of a normal one. Of course, you can only shine it on something if you can see the target behind the wall, like via x-ray or thermals. Mesons don't count, however.
If one tries to jam a crystal into a pointer with a tier 1/2 laser (or a tier 1/2 laser in a pointer with an installed crystal), _something_ will get teleported, crushing the crystal.
You can uninstall the crystal with wirecutters or a hemostat. The pointer will _hint_ on closer examination (`examine_more`) at a possibility of a crystal being installed if you upgrade the laser (different messages for tier 1/2/3,4).
Removes one stupid 1% increase for a recharge chance per process tick if your laser was in a full recharge state because it was insignificant and irrelevant.

i've had a branch for this for almost 9 months and i was always laying it off for some day later. today i just completely fucked the branch. whoops. i'm not even sure at this point what else did i fix while here, double whoops

## Why It's Good For The Game
Closes #45834 - Canonizes a bug into a feature.
Fixes #77003 - lol
Cleaner code, possibly more robust even.
Seeing the remaining charge was not available at all and the only hint was when you tried shining the pointer on something. That sucks.

## Changelog
:cl:
add: you can upgrade laser pointers with a bluespace crystal to let them shine through walls at double the power cost, if the laser in the pointer is of tier 3 or higher.
qol: laser pointer charge can be seen by examining it
fix: fixed laser pointers luring dead cats when shone upon
code: laser pointer code cleaned up a tad
/:cl:

Sealed101 committed:
my old branch keeled over
- dmdoc laser pointer, improve code
- shining through walls is now a feature - but you must install a bs crystal
- need a laser of tier 3 and up to fit a crystal
- lower tier lasers fumble installation by teleporting you, the pointer or the laser
- stored energy is available on examine
recharge lock doesn't improve the chance to recharge per process tick by 1 (this is hilariously insignificant why is it here)

Sealed101 commented on code/game/objects/items/devices/laserpointer.dm:
```diff
@@ -44,22 +52,78 @@
+				do_teleport(atom_to_teleport, get_turf(src), jammed_crystal.blink_range, asoundin = 'sound/effects/phasein.ogg', channel = TELEPORT_CHANNEL_BLUESPACE)
+				jammed_crystal.use_tool(src, user, amount = 1) //use only one if we were installing from a stack of crystals
+			return
+		if(!jammed_crystal.use_tool(src, user, amount = 1))
+			return
+		crystal_lens = new(src) //this could probably convert artificial crystals into pure ones, hmm
```
i did not come up with a good solution for this then and i don't think i will come up with one now, pls halp

carlarctg replied:
feature, it teleports a normal crystal in. Bluespace

Sealed101 replied:
oh #77092 could come in useful for solving this by making `fast_split_stack()` global, good to know

Sealed101 commented on code/game/objects/items/devices/laserpointer.dm:
```diff
@@ -68,103 +132,133 @@
 	icon_state = "pointer_[pointer_icon_state]"
+
+	//setup pointer blip
 	var/mutable_appearance/laser = mutable_appearance('icons/obj/weapons/guns/projectiles.dmi', pointer_icon_state, 10)
```
i also remember trying to figure out why the pointer image was not layering over anything except tiles and atmos devices, it's still a problem even with the `mutable_appearance` thing

Sealed101 reviewed (commented):
two things in need of further fixing

may also be a good idea to check for death/status on other affected things

ATH1909 commented:
can you un-upgrade a laser pointer that's been augmented with a bluespace crystal? does it work if you can see your target through cameras, provided they're in range?

Sealed101 commented:
> can you un-upgrade a laser pointer that's been augmented with a bluespace crystal? 

wirecutter/hemostat to remove the crystal

> does it work if you can see your target through cameras, provided they're in range?

i vaguely remember testing it during the fucking 9 months that i've been sitting on this, but i'm not sure. you can just pop into a live server now and test it out though, it should be working (or not working) similarly to that

Sealed101 commented on code/game/objects/items/devices/laserpointer.dm:
```diff
@@ -68,103 +132,133 @@
 		to_chat(user, span_warning("Your fingers can't press the button!"))
 		return
+
+	if(!(user in (viewers(7, target)))) //check if we can see the target (xray and thermals can give us vision through obstacles)
+		return
+	if(!(user in (view(7, target)))) //check if we are visible from the target's PoV
+		if(!crystal_lens)
+			to_chat(user, span_warning("You can't point with [src] through walls!"))
+			return
+		if(!((user.sight & SEE_OBJS) || (user.sight & SEE_MOBS))) //only let it work if we have xray or thermals. mesons don't count because they are easier to get.
+			to_chat(user, span_notice("You can't quite make out your target and you fail to shine at it."))
+			return
```
i tried to make it so you need to have xray or thermals to actually be able to point at stuff behind walls, so meson vision won't be plausible. however, combining meson vision with thermal vision (i.e. thermal mutation + meson goggles) does allow targeting any atom, though you won't see the blip. i kinda feel like it's fine, since that meson-thermals combo is rare to come by and in case of the mutation combo it's only available for a short window of time

carlarctg commented on code/game/objects/items/devices/laserpointer.dm:
```diff
@@ -44,22 +52,78 @@
+		//the crystal we're about to install
+		var/obj/item/stack/ore/bluespace_crystal/jammed_crystal = attack_item
+		if(diode && diode.rating < 3) //only lasers of tier 3 and up can house a lens
+			to_chat(user, span_warning("You try to jam \the [jammed_crystal.name] in front of the diode, but it's a bad fit!"))
```
```suggestion
			to_chat(user, span_warning("You try to jam \the [jammed_crystal.name] in front of the diode, but the diode is taking up all the space!"))
```

carlarctg commented:
I'm kind of impressed at how feature-bloated you made the tiny addition of 'laser pointers are xray compatible with bluespace crystal'.

I don't like the laser bluespace thing being based on size at all, it makes no sense (all the lasers are the same size both in sprites and in game), and is unintuitive. Personally I would reflavor it along the lines of 'The laser is not fine or precise enough to pass through the bluespace crystal properly!' and then teleport the laser to a random onscreen tile or something. 

Hell you could even have that be the thing - the more upgraded the laser is, the less the laser is teleported randomly around its location.

This is a cool PR though. Now I want telecrystal laser pointers...

Sealed101 commented:
this isn't that bloated. 
it could be more bloated if i sat my ass down back then and went through with gagsifying the laser and actually implementing the telecrystal upgrade. but now i just don't want to hold this back any longer
flavoring it around diode fidelity instead of diode size also doesn't allow for the teleportation fumble to make sense. not even with the "it's bluespace" card

Sealed101 committed:
more stat/status checks

Sealed101 committed:
Merge branch 'master' into laser-pointer-stuff

Sealed101 committed:
Merge branch 'master' into laser-pointer-stuff

Sealed101 committed:
fast_split_stack for crystal installing

Sealed101 committed:
we had a split_stack this whole time lmao

LemonInTheDark commented on code/game/objects/items/devices/laserpointer.dm:
```diff
@@ -5,19 +5,27 @@
+	custom_materials = list(/datum/material/iron = SMALL_MATERIAL_AMOUNT * 5, /datum/material/glass = SMALL_MATERIAL_AMOUNT * 5)
 	w_class = WEIGHT_CLASS_SMALL
-	var/turf/pointer_loc
+	///Currently stored blulespace crystal, if any. Required to use the pointer through walls
+	var/obj/item/stack/ore/bluespace_crystal/crystal_lens
+	///Currently stored micro-laser diode
+	var/obj/item/stock_parts/micro_laser/diode
```
what happens if one of these was suddenly Move()d out of the pointer, would the ref remain?

LemonInTheDark commented on code/game/objects/items/devices/laserpointer.dm:
```diff
@@ -68,103 +136,140 @@
 		to_chat(user, span_warning("Your fingers can't press the button!"))
 		return
+
+	if(!(user in (viewers(7, target)))) //check if we can see the target (xray and thermals can give us vision through obstacles)
+		return
+	if(!(user in (view(7, target)))) //check if we are visible from the target's PoV
```
why do we do both this sucks

LemonInTheDark reviewed (changes requested):
I'm fine with the sight stuff that case is so rare it doesn't matter

Sealed101 committed:
it was using target's turf for the pointer overlay
lmao i'm blind (possibly because of all the laser pointers)
also handles destroy and atom exit

Sealed101 committed:
turns out we don't quite need this one
kind of ehhh at the failure messages but they kind of work

Sealed101 committed:
review bits
sometimes i wonder if my brain rot will get the better of me

Sealed101 committed:
tab

Co-authored-by: Jacquerel <hnevard@gmail.com>

Ghommie commented on code/game/objects/items/devices/laserpointer.dm:
```diff
@@ -44,128 +64,226 @@
 		to_chat(user, span_warning("Your fingers can't press the button!"))
 		return
+
+	if(!(user in (view(7, target)))) //check if we are visible from the target's PoV
```
Beside the fact this checks a hardcoded value of 7 rather than the client view, doesn't this mean a mob with thermal or x-ray vision could be able to shine a laser light everywhere on the station with the help or a (advanced) camera console?

Sealed101 replied:
nope, they point at a screen object that tgui shows in its window

Ghommie commented on code/game/objects/items/devices/laserpointer.dm:
```diff
@@ -44,128 +64,226 @@
+		if(diode && diode.rating < 3) //only lasers of tier 3 and up can house a lens
+			to_chat(user, span_warning("You try to jam \the [crystal_stack.name] in front of the diode, but it's a bad fit!"))
+			playsound(src, 'sound/machines/airlock_alien_prying.ogg', 20)
+			if(do_after(user, 2 SECONDS, src))
```
I think these do_after actions and what follows could share a common proc.

Sealed101 replied:
i'd agree if i made them equal in possible results which i didn't nor do i intend to. one's teleporting either the diode or the user, the other's teleporting either the pointer or the user; writing out more typechecks for differing `to_chat()` messages would probably just make it more of a mess than it should be

Sealed101 committed:
i stand corrected
station camera consoles could actually let you point at objects(not turfs) too far away given that you had xray/thermals
advanced camera consoles seem to be fucky in some other way by affecting user's sight so they don't let you shine through walls at all lmao
so let's just plug a range check before checking sight
also makes the pointer max range a var because why not - defaults to 7 bc i don't want to touch sight ranges and adjacent things
plus uhh did you ever try pointing a laser to a far away thing? yeah you won't have precision
